File: src/main/cpp/g0101_0200/s0138_copy_list_with_random_pointer/Solution.cpp
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Hash_Table #Linked_List
// #Programming_Skills_II_Day_14 #Udemy_Linked_List #Big_O_Time_O(N)_Space_O(N)
// #2024_05_27_Time_5_ms_(65.69%)_Space_12.9_MB_(73.88%)

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == nullptr) {
            return nullptr;
        }
        Node* curr = head;
        // First pass to create cloned nodes and insert them between original nodes
        while (curr != nullptr) {
            Node* clonedNode = new Node(curr->val, curr->next, nullptr);
            curr->next = clonedNode;
            curr = clonedNode->next;
        }
        // Second pass to update random pointers of cloned nodes
        curr = head;
        while (curr != nullptr) {
            if (curr->random != nullptr) {
                curr->next->random = curr->random->next;
            }
            curr = curr->next->next;
        }
        // Third pass to separate original and cloned nodes
        curr = head;
        Node* newHead = curr->next;
        while (curr != nullptr) {
            Node* clonedNode = curr->next;
            curr->next = clonedNode->next;
            if (clonedNode->next != nullptr) {
                clonedNode->next = clonedNode->next->next;
            }
            curr = curr->next;
        }
        return newHead;
    }
};

File: src/main/python/g0101_0200/s0138_copy_list_with_random_pointer/Solution.py
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Hash_Table #Linked_List
# #Programming_Skills_II_Day_14 #Udemy_Linked_List #Big_O_Time_O(N)_Space_O(N)
# #2024_06_09_Time_36_ms_(72.10%)_Space_17.3_MB_(95.85%)

"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if head is None:
            return None
        
        # First pass: create the cloned nodes and insert them right after the original nodes
        curr = head
        while curr:
            cloned_node = Node(curr.val)
            cloned_node.next = curr.next
            curr.next = cloned_node
            curr = cloned_node.next
        
        # Second pass: update the random pointers of the cloned nodes
        curr = head
        while curr:
            if curr.random:
                curr.next.random = curr.random.next
            curr = curr.next.next
        
        # Third pass: separate the cloned list from the original list
        curr = head
        new_head = head.next
        while curr:
            cloned_node = curr.next
            curr.next = cloned_node.next
            if cloned_node.next:
                cloned_node.next = cloned_node.next.next
            curr = curr.next
        
        return new_head
